### ISSUE TITLE
11 preview5 - hangs on tooltip close on macos(not tested on other os)

### ISSUE THREAD

**Issue by hellozyemlya** [AvaloniaUI/Avalonia#10392]
**Describe the bug**
Application hangs when lot of tooltips open-close happens. Sometimes it just crashing, was not able to catch stacktrace or crash report.

**To Reproduce**
Create lot of elements with tooltips, move mouse here and there, wait until app hangs

**Expected behavior**
No hang

**Desktop (please complete the following information):**

- OS: mac os m1
- Version: 11 preview5

**Additional context**
Stacktrace
Monitor.Wait()at /Users/eugenechekanskiy/Library/Application Support/JetBrains/Rider2022.3/resharper-host/SourcesCache/bcae882191178c8bf44c9add4cdeffa874a7afe629e3e56e9280f5896e953425/Monitor.CoreCLR.cs:line 159
ManualResetEventSlim.Wait()
Task.SpinThenBlockingWait()
Task.InternalWaitCore()
Task.Wait()
Task.Wait()
CompositingRenderer.Dispose()at /_/src/Avalonia.Base/Rendering/Composition/CompositingRenderer.cs:line 324
TopLevel.HandleClosed()
WindowBase.HandleClosed()
WindowBaseImpl.WindowBaseEvents.Avalonia.Native.Interop.IAvnWindowBaseEvents.Closed()
__MicroComIAvnWindowBaseEventsVTable.Closed()
[Managed to Native Transition]
__MicroComIAvnWindowBaseProxy.Close()
WindowBaseImpl.Dispose()
PopupRoot.Dispose()
ToolTip.Close()
ToolTip.IsOpenChanged()
AnonymousObserver<AvaloniaPropertyChangedEventArgs<bool>>.OnNext()
LightweightObservableBase<AvaloniaPropertyChangedEventArgs<bool>>.PublishNext()
LightweightSubject<AvaloniaPropertyChangedEventArgs<bool>>.OnNext()
AvaloniaProperty<bool>.NotifyChanged()
AvaloniaObject.RaisePropertyChanged<bool>()
EffectiveValue<bool>.SetAndRaiseCore()
EffectiveValue<bool>.SetLocalValueAndRaise()
ValueStore.SetValue<bool>()
AvaloniaObject.SetValue<bool>()
ToolTip.SetIsOpen()
ToolTipService.Close()
ToolTipService.ControlPointerExited()
Interactive.<AddHandler>g__InvokeAdapter|4_0<Avalonia.Input.PointerEventArgs>()
Interactive.<>c__4<PointerEventArgs>.<AddHandler>b__4_1()
EventRoute.RaiseEventImpl()
EventRoute.RaiseEvent()
Interactive.RaiseEvent()
PointerOverPreProcessor.SetPointerOverToElement()
PointerOverPreProcessor.SetPointerOver()
PointerOverPreProcessor.SceneInvalidated()
TopLevel.SceneInvalidated()
async CompositingRenderer.TriggerSceneInvalidatedOnBatchCompletion()
AsyncTaskMethodBuilder<VoidTaskResult>.AsyncStateMachineBox<CompositingRenderer.<TriggerSceneInvalidatedOnBatchCompletion>d__29>.ExecutionContextCallback()
ExecutionContext.RunInternal()
AsyncTaskMethodBuilder<VoidTaskResult>.AsyncStateMachineBox<CompositingRenderer.<TriggerSceneInvalidatedOnBatchCompletion>d__29>.MoveNext()
AsyncTaskMethodBuilder<VoidTaskResult>.AsyncStateMachineBox<CompositingRenderer.<TriggerSceneInvalidatedOnBatchCompletion>d__29>.MoveNext()
AwaitTaskContinuation.<>c.<.cctor>b__17_0()
AwaitTaskContinuation.RunCallback()
SynchronizationContextAwaitTaskContinuation.Run()
Task.RunContinuations()
Task.FinishContinuations()
Task<int>.TrySetResult()
TaskCompletionSource<int>.TrySetResult()
Batch.Complete()
ServerCompositor.CompletePendingBatches()
ServerCompositor.RenderCore()
ServerCompositor.Render()
CompositionTarget.ImmediateUIThreadRender()
CompositingRenderer.Paint()
TopLevel.HandlePaint()
WindowBaseImpl.WindowBaseEvents.Avalonia.Native.Interop.IAvnWindowBaseEvents.Paint()
__MicroComIAvnWindowBaseEventsVTable.Paint()
[Managed to Native Transition]
__MicroComIAvnPlatformThreadingInterfaceProxy.RunLoop()
PlatformThreadingInterface.RunLoop()
Dispatcher.MainLoop()
ClassicDesktopStyleApplicationLifetime.Start()
ClassicDesktopStyleApplicationLifetimeExtensions.StartWithClassicDesktopLifetime()
Program.Main()

**Comment by hellozyemlya** [comment]
Unfortunately, was not able to create minimal application that reproducing issue.
Still some details here - looks like ServerCompositor.Render blocked on lock (_lock) call. Still not sure who is currently holding lock... See stack trace below.

Monitor.Enter()at line 49
ServerCompositor.Render()at /Users/eugenechekanskiy/Develop/lego/avaloniafork/build/Avalonia/src/Avalonia.Base/Rendering/Composition/Server/ServerCompositor.cs:line 136
RenderLoop.TimerTick()
DefaultRenderTimer.InternalTick()
DefaultRenderTimer.<>c__DisplayClass14_0.<StartCore>b__0()
StandardRuntimePlatform.<>c__DisplayClass0_0.<StartSystemTimer>b__0()
TimerQueueTimer.<>c.<.cctor>b__27_0()
ExecutionContext.RunInternal()
TimerQueueTimer.CallCallback()
TimerQueueTimer.Fire()
TimerQueue.FireNextTimers()
TimerQueue.AppDomainTimerCallback()
UnmanagedThreadPoolWorkItem.System.Threading.IThreadPoolWorkItem.Execute()
[Native to Managed Transition]
[Managed to Native Transition]
ThreadPoolWorkQueue.Dispatch()
PortableThreadPool.WorkerThread.WorkerThreadStart()
Thread.StartCallback()
[Native to Managed Transition]

**Comment by hellozyemlya** [comment]
Not reproduced in same application on windows.

**Comment by hellozyemlya** [comment]
Probability of getting this lock much higher if tooltip delay is 0.

**Comment by hellozyemlya** [comment]
Here is a repository with reproduce https://github.com/hellozyemlya/AvaloniaTooltipCloseDeadlock.
Reproduce is tricky: you need move mouse with different speed on top of toggles. For me it usually works like this - move fast, move slowly, move away from window and move back.

**Comment by hellozyemlya** [comment]
If repository will not be enough, can provide published build as well.

**Comment by maxkatz6** [comment]
Seems to be a duplicate of
https://github.com/AvaloniaUI/Avalonia/issues/9890

**Comment by hellozyemlya** [comment]
@maxkatz6 Luck we have it reproduced on mac os. Unluck - issue seems to be serious one :( Is there some way to disable composition render or somehow to tweak it and ensure it does not hangs?

**Comment by maxkatz6** [comment]
@hellozyemlya preview versions are expected to have serious problems.

If you can stably reproduce this issue, can you check these comments?
https://github.com/AvaloniaUI/Avalonia/issues/9890#issuecomment-1373347215
https://github.com/AvaloniaUI/Avalonia/issues/9890#issuecomment-1373976085

**Comment by hellozyemlya** [comment]
@maxkatz6 @kekekeks Stacks attached below. Also, trace logs are clear. As well as setting break on any exceptions does not affect anything, there is no exceptions. Let me know if any other assistance required in order to debug it. Fortunately I can reproduce it relatively easy.
```
 @3822455

 @3822457

.NET ThreadPool Gate @3822495
	WaitHandle.WaitOneNoCheck()
	WaitHandle.WaitOne()
	PortableThreadPool.GateThread.GateThreadStart()
	Thread.StartCallback()
	[Native to Managed Transition]

.NET ThreadPool Worker @3822494
	Monitor.Enter()at /Users/eugenechekanskiy/Library/Application Support/JetBrains/Rider2022.3/resharper-host/SourcesCache/bcae882191178c8bf44c9add4cdeffa874a7afe629e3e56e9280f5896e953425/Monitor.CoreCLR.cs:line 49
	ServerCompositor.Render()at /_/src/Avalonia.Base/Rendering/Composition/Server/ServerCompositor.cs:line 136
	RenderLoop.TimerTick()
	DefaultRenderTimer.InternalTick()
	DefaultRenderTimer.<>c__DisplayClass14_0.<StartCore>b__0()
	StandardRuntimePlatform.<>c__DisplayClass0_0.<StartSystemTimer>b__0()
	TimerQueueTimer.<>c.<.cctor>b__27_0()
	ExecutionContext.RunInternal()
	TimerQueueTimer.CallCallback()
	TimerQueueTimer.Fire()
	TimerQueue.FireNextTimers()
	TimerQueue.AppDomainTimerCallback()
	UnmanagedThreadPoolWorkItem.System.Threading.IThreadPoolWorkItem.Execute()
	[Native to Managed Transition]
	[Managed to Native Transition]
	ThreadPoolWorkQueue.Dispatch()
	PortableThreadPool.WorkerThread.WorkerThreadStart()
	Thread.StartCallback()
	[Native to Managed Transition]

.NET ThreadPool Worker @3822542
	[Native frame]
	LowLevelLifoSemaphore.WaitForSignal()
	LowLevelLifoSemaphore.Wait()
	PortableThreadPool.WorkerThread.WorkerThreadStart()
	Thread.StartCallback()
	[Native to Managed Transition]

.NET ThreadPool Worker @3822504
	[Native frame]
	LowLevelLifoSemaphore.WaitForSignal()
	LowLevelLifoSemaphore.Wait()
	PortableThreadPool.WorkerThread.WorkerThreadStart()
	Thread.StartCallback()
	[Native to Managed Transition]

.NET ThreadPool Worker @3822503
	[Native frame]
	LowLevelLifoSemaphore.WaitForSignal()
	LowLevelLifoSemaphore.Wait()
	PortableThreadPool.WorkerThread.WorkerThreadStart()
	Thread.StartCallback()
	[Native to Managed Transition]

Main Thread @3822431
	Monitor.Wait()at /Users/eugenechekanskiy/Library/Application Support/JetBrains/Rider2022.3/resharper-host/SourcesCache/bcae882191178c8bf44c9add4cdeffa874a7afe629e3e56e9280f5896e953425/Monitor.CoreCLR.cs:line 159
	ManualResetEventSlim.Wait()
	Task.SpinThenBlockingWait()
	Task.InternalWaitCore()
	Task.Wait()
	Task.Wait()
	CompositingRenderer.Dispose()at /_/src/Avalonia.Base/Rendering/Composition/CompositingRenderer.cs:line 324
	TopLevel.HandleClosed()
	WindowBase.HandleClosed()
	WindowBaseImpl.WindowBaseEvents.Avalonia.Native.Interop.IAvnWindowBaseEvents.Closed()
	__MicroComIAvnWindowBaseEventsVTable.Closed()
	[Managed to Native Transition]
	__MicroComIAvnWindowBaseProxy.Close()
	WindowBaseImpl.Dispose()
	PopupRoot.Dispose()
	ToolTip.Close()
	ToolTip.IsOpenChanged()
	AnonymousObserver<AvaloniaPropertyChangedEventArgs<bool>>.OnNext()
	LightweightObservableBase<AvaloniaPropertyChangedEventArgs<bool>>.PublishNext()
	LightweightSubject<AvaloniaPropertyChangedEventArgs<bool>>.OnNext()
	AvaloniaProperty<bool>.NotifyChanged()
	AvaloniaObject.RaisePropertyChanged<bool>()
	EffectiveValue<bool>.SetAndRaiseCore()
	EffectiveValue<bool>.SetLocalValueAndRaise()
	ValueStore.SetValue<bool>()
	AvaloniaObject.SetValue<bool>()
	ToolTip.SetIsOpen()
	ToolTipService.Close()
	ToolTipService.ControlPointerExited()
	Interactive.<AddHandler>g__InvokeAdapter|4_0<Avalonia.Input.PointerEventArgs>()
	Interactive.<>c__4<PointerEventArgs>.<AddHandler>b__4_1()
	EventRoute.RaiseEventImpl()
	EventRoute.RaiseEvent()
	Interactive.RaiseEvent()
	PointerOverPreProcessor.SetPointerOverToElement()
	PointerOverPreProcessor.SetPointerOver()
	PointerOverPreProcessor.SceneInvalidated()
	TopLevel.SceneInvalidated()
	async CompositingRenderer.TriggerSceneInvalidatedOnBatchCompletion()
	AsyncTaskMethodBuilder<VoidTaskResult>.AsyncStateMachineBox<CompositingRenderer.<TriggerSceneInvalidatedOnBatchCompletion>d__29>.ExecutionContextCallback()
	ExecutionContext.RunInternal()
	AsyncTaskMethodBuilder<VoidTaskResult>.AsyncStateMachineBox<CompositingRenderer.<TriggerSceneInvalidatedOnBatchCompletion>d__29>.MoveNext()
	AsyncTaskMethodBuilder<VoidTaskResult>.AsyncStateMachineBox<CompositingRenderer.<TriggerSceneInvalidatedOnBatchCompletion>d__29>.MoveNext()
	AwaitTaskContinuation.<>c.<.cctor>b__17_0()
	AwaitTaskContinuation.RunCallback()
	SynchronizationContextAwaitTaskContinuation.Run()
	Task.RunContinuations()
	Task.FinishContinuations()
	Task<int>.TrySetResult()
	TaskCompletionSource<int>.TrySetResult()
	Batch.Complete()
	ServerCompositor.CompletePendingBatches()
	ServerCompositor.RenderCore()
	ServerCompositor.Render()
	CompositionTarget.ImmediateUIThreadRender()
	CompositingRenderer.Paint()
	TopLevel.HandlePaint()
	WindowBaseImpl.WindowBaseEvents.Avalonia.Native.Interop.IAvnWindowBaseEvents.Paint()
	__MicroComIAvnWindowBaseEventsVTable.Paint()
	[Managed to Native Transition]
	__MicroComIAvnPlatformThreadingInterfaceProxy.RunLoop()
	PlatformThreadingInterface.RunLoop()
	Dispatcher.MainLoop()
	ClassicDesktopStyleApplicationLifetime.Start()
	ClassicDesktopStyleApplicationLifetimeExtensions.StartWithClassicDesktopLifetime()
	Program.Main()

Worker thread @3822493


 @3822455

 @3822457

.NET ThreadPool Gate @3822495
	WaitHandle.WaitOneNoCheck()
	WaitHandle.WaitOne()
	PortableThreadPool.GateThread.GateThreadStart()
	Thread.StartCallback()
	[Native to Managed Transition]

.NET ThreadPool Worker @3822494
	Monitor.Enter()at /Users/eugenechekanskiy/Library/Application Support/JetBrains/Rider2022.3/resharper-host/SourcesCache/bcae882191178c8bf44c9add4cdeffa874a7afe629e3e56e9280f5896e953425/Monitor.CoreCLR.cs:line 49
	ServerCompositor.Render()at /_/src/Avalonia.Base/Rendering/Composition/Server/ServerCompositor.cs:line 136
	RenderLoop.TimerTick()
	DefaultRenderTimer.InternalTick()
	DefaultRenderTimer.<>c__DisplayClass14_0.<StartCore>b__0()
	StandardRuntimePlatform.<>c__DisplayClass0_0.<StartSystemTimer>b__0()
	TimerQueueTimer.<>c.<.cctor>b__27_0()
	ExecutionContext.RunInternal()
	TimerQueueTimer.CallCallback()
	TimerQueueTimer.Fire()
	TimerQueue.FireNextTimers()
	TimerQueue.AppDomainTimerCallback()
	UnmanagedThreadPoolWorkItem.System.Threading.IThreadPoolWorkItem.Execute()
	[Native to Managed Transition]
	[Managed to Native Transition]
	ThreadPoolWorkQueue.Dispatch()
	PortableThreadPool.WorkerThread.WorkerThreadStart()
	Thread.StartCallback()
	[Native to Managed Transition]

.NET ThreadPool Worker @3822542
	[Native frame]
	LowLevelLifoSemaphore.WaitForSignal()
	LowLevelLifoSemaphore.Wait()
	PortableThreadPool.WorkerThread.WorkerThreadStart()
	Thread.StartCallback()
	[Native to Managed Transition]

.NET ThreadPool Worker @3822504
	[Native frame]
	LowLevelLifoSemaphore.WaitForSignal()
	LowLevelLifoSemaphore.Wait()
	PortableThreadPool.WorkerThread.WorkerThreadStart()
	Thread.StartCallback()
	[Native to Managed Transition]

.NET ThreadPool Worker @3822503
	[Native frame]
	LowLevelLifoSemaphore.WaitForSignal()
	LowLevelLifoSemaphore.Wait()
	PortableThreadPool.WorkerThread.WorkerThreadStart()
	Thread.StartCallback()
	[Native to Managed Transition]

Main Thread @3822431
	Monitor.Wait()at /Users/eugenechekanskiy/Library/Application Support/JetBrains/Rider2022.3/resharper-host/SourcesCache/bcae882191178c8bf44c9add4cdeffa874a7afe629e3e56e9280f5896e953425/Monitor.CoreCLR.cs:line 159
	ManualResetEventSlim.Wait()
	Task.SpinThenBlockingWait()
	Task.InternalWaitCore()
	Task.Wait()
	Task.Wait()
	CompositingRenderer.Dispose()at /_/src/Avalonia.Base/Rendering/Composition/CompositingRenderer.cs:line 324
	TopLevel.HandleClosed()
	WindowBase.HandleClosed()
	WindowBaseImpl.WindowBaseEvents.Avalonia.Native.Interop.IAvnWindowBaseEvents.Closed()
	__MicroComIAvnWindowBaseEventsVTable.Closed()
	[Managed to Native Transition]
	__MicroComIAvnWindowBaseProxy.Close()
	WindowBaseImpl.Dispose()
	PopupRoot.Dispose()
	ToolTip.Close()
	ToolTip.IsOpenChanged()
	AnonymousObserver<AvaloniaPropertyChangedEventArgs<bool>>.OnNext()
	LightweightObservableBase<AvaloniaPropertyChangedEventArgs<bool>>.PublishNext()
	LightweightSubject<AvaloniaPropertyChangedEventArgs<bool>>.OnNext()
	AvaloniaProperty<bool>.NotifyChanged()
	AvaloniaObject.RaisePropertyChanged<bool>()
	EffectiveValue<bool>.SetAndRaiseCore()
	EffectiveValue<bool>.SetLocalValueAndRaise()
	ValueStore.SetValue<bool>()
	AvaloniaObject.SetValue<bool>()
	ToolTip.SetIsOpen()
	ToolTipService.Close()
	ToolTipService.ControlPointerExited()
	Interactive.<AddHandler>g__InvokeAdapter|4_0<Avalonia.Input.PointerEventArgs>()
	Interactive.<>c__4<PointerEventArgs>.<AddHandler>b__4_1()
	EventRoute.RaiseEventImpl()
	EventRoute.RaiseEvent()
	Interactive.RaiseEvent()
	PointerOverPreProcessor.SetPointerOverToElement()
	PointerOverPreProcessor.SetPointerOver()
	PointerOverPreProcessor.SceneInvalidated()
	TopLevel.SceneInvalidated()
	async CompositingRenderer.TriggerSceneInvalidatedOnBatchCompletion()
	AsyncTaskMethodBuilder<VoidTaskResult>.AsyncStateMachineBox<CompositingRenderer.<TriggerSceneInvalidatedOnBatchCompletion>d__29>.ExecutionContextCallback()
	ExecutionContext.RunInternal()
	AsyncTaskMethodBuilder<VoidTaskResult>.AsyncStateMachineBox<CompositingRenderer.<TriggerSceneInvalidatedOnBatchCompletion>d__29>.MoveNext()
	AsyncTaskMethodBuilder<VoidTaskResult>.AsyncStateMachineBox<CompositingRenderer.<TriggerSceneInvalidatedOnBatchCompletion>d__29>.MoveNext()
	AwaitTaskContinuation.<>c.<.cctor>b__17_0()
	AwaitTaskContinuation.RunCallback()
	SynchronizationContextAwaitTaskContinuation.Run()
	Task.RunContinuations()
	Task.FinishContinuations()
	Task<int>.TrySetResult()
	TaskCompletionSource<int>.TrySetResult()
	Batch.Complete()
	ServerCompositor.CompletePendingBatches()
	ServerCompositor.RenderCore()
	ServerCompositor.Render()
	CompositionTarget.ImmediateUIThreadRender()
	CompositingRenderer.Paint()
	TopLevel.HandlePaint()
	WindowBaseImpl.WindowBaseEvents.Avalonia.Native.Interop.IAvnWindowBaseEvents.Paint()
	__MicroComIAvnWindowBaseEventsVTable.Paint()
	[Managed to Native Transition]
	__MicroComIAvnPlatformThreadingInterfaceProxy.RunLoop()
	PlatformThreadingInterface.RunLoop()
	Dispatcher.MainLoop()
	ClassicDesktopStyleApplicationLifetime.Start()
	ClassicDesktopStyleApplicationLifetimeExtensions.StartWithClassicDesktopLifetime()
	Program.Main()

Worker thread @3822493
```

**Comment by hellozyemlya** [comment]
It seems like thread with batch somehow died and lock keeps unreleased... Is there any way to check thread id that hold lock currently?

**Comment by maxkatz6** [comment]
Might be fixed in master.
